src/io/fasta.rs

```diff
@@ -252,12 +366,73 @@ struct IndexRecord {
 
 
 /// A sequence record returned by the FASTA index.
+#[derive(Debug, PartialEq)]
 pub struct Sequence {
     pub name: String,
     pub len: u64,
 }
 
 
+pub struct IndexedReaderIterator<'a, R: io::Read + io::Seek + 'a> {
+    reader: &'a mut IndexedReader<R>,
+    record: IndexRecord,
+    bases_left: u64,
+    line_offset: u64,
+    buf: Vec<u8>,
+    buf_idx: usize,
+}
+
+
+impl<'a, R: io::Read + io::Seek + 'a> IndexedReaderIterator<'a, R> {
+    fn fill_buffer(&mut self) -> io::Result<()> {
+        assert!(self.bases_left > 0);
+
+        self.buf.clear();
+        let bases_to_read = min(self.buf.capacity() as u64, self.bases_left);
+
+        while self.buf.is_empty() {
```

Comment:
is_empty() can be true the following number of times:

1. The common case, where line_offset < line_bases.
2. The uncommon case, where line_bases <= line_offset.
3. Files with silly amounts of whitespace can loop any amount of times.

Two or more loops can happen if the last read ended at or inside the whitespace, on lines longer than self.buf.capacity(), in which case the 'read_line' call only consumes whitespace. More than two loops would, however, require some pretty silly amounts of whitespace in your file, so that would be limited to poorly formatted files (unintentionally or malicious).